### ISSUE TITLE
python code

### ISSUE THREAD

**Issue by BudiPranata** [BudiPranata/webddream#1]
import pandas as pd
import numpy as np
import scipy.stats
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler
from sklearn.svm import SVR
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error, mean_absolute_error, mean_absolute_percentage_error
import warnings
warnings.filterwarnings('ignore')

# ==========================================
# 1. FUNGSI DIEBOLD-MARIANO TEST (STATISTIK)
# ==========================================
def diebold_mariano_test(actual, pred1, pred2):
    e1 = np.array(actual) - np.array(pred1)
    e2 = np.array(actual) - np.array(pred2)

    # Menggunakan Squared Error (sesuai RMSE)
    d = (e1 ** 2) - (e2 ** 2)
    mean_d = np.mean(d)
    var_d = np.var(d, ddof=1)
    n = len(d)

    if var_d == 0:
        return 0, 1.0

    dm_stat = mean_d / np.sqrt(var_d / n)
    p_value = 2 * (1 - scipy.stats.norm.cdf(abs(dm_stat)))
    return dm_stat, p_value

# ==========================================
# 2. FUNGSI RECURSIVE KALMAN FILTER
# ==========================================
def recursive_kalman_filter(y_train, y_test):
    # Inisialisasi status awal menggunakan data terakhir dari train set
    x_est = y_train[-1]
    p_est = 1.0
    Q = 1e-4  # Process noise variance
    R = 1e-2  # Measurement noise variance

    predictions = []

    for measurement in y_test:
        # 1. Tahap Prediksi (Time Update)
        x_pred = x_est
        p_pred = p_est + Q
        predictions.append(x_pred)

        # 2. Tahap Koreksi (Measurement Update) menggunakan observasi aktual
        K = p_pred / (p_pred + R)
        x_est = x_pred + K * (measurement - x_pred)
        p_est = (1 - K) * p_pred

    return np.array(predictions)

# ==========================================
# 3. MUAT & PRA-PEMROSESAN DATA
# ==========================================
print("Memuat dataset...")
file_path = 'profit.csv'
try:
    df = pd.read_csv(file_path, thousands='.')
except FileNotFoundError:
    print(f"Error: File '{file_path}' tidak ditemukan. Pastikan path sudah benar.")
    exit()

# Fitur operasional (kolom ke-2 sampai sebelum terakhir)
X = df.iloc[:, 1:-1].values
# Target (Profit) pada kolom terakhir
y = df.iloc[:, -1].values

# Normalisasi data (Min-Max Scaling untuk stabilitas perhitungan SVM & KF)
scaler_X = MinMaxScaler()
scaler_y = MinMaxScaler()

X_scaled = scaler_X.fit_transform(X)
y_scaled = scaler_y.fit_transform(y.reshape(-1, 1)).flatten()

# ==========================================
# 4. INISIALISASI VARIABEL EVALUASI METRIK
# ==========================================
split_ratios = [0.70, 0.75, 0.80, 0.85, 0.90]
n_total = len(df)

# Dictionary untuk menyimpan hasil metrik seluruh skenario
eval_results = {
    'Ratio': [],
    'Naive_RMSE': [], 'KF_RMSE': [], 'SVM_RMSE': [],
    'Naive_MAE': [], 'KF_MAE': [], 'SVM_MAE': [],
    'Naive_MAPE': [], 'KF_MAPE': [], 'SVM_MAPE': []
}

print("\nMemulai Eksperimen Time-Series Cross-Validation...\n")

# ==========================================
# 5. SKENARIO WALK-FORWARD VALIDATION (LOOP)
# ==========================================
for ratio in split_ratios:
    ratio_str = f"{int(ratio*100)}:{int((1-ratio)*100)}"
    print(f"{'='*60}")
    print(f"SKENARIO PEMBAGIAN DATA: {ratio_str}")
    print(f"{'='*60}")

    train_size = int(n_total * ratio)

    # Pembagian kronologis yang mempertahankan struktur deret waktu
    X_train, X_test = X_scaled[:train_size], X_scaled[train_size:]
    y_train, y_test = y_scaled[:train_size], y_scaled[train_size:]

    # A. NAIVE BASELINE MODEL
    y_pred_naive_scaled = np.roll(y_test, shift=1)
    y_pred_naive_scaled[0] = y_train[-1]

    # B. RECURSIVE KALMAN FILTER
    y_pred_kf_scaled = recursive_kalman_filter(y_train, y_test)

    # C. SUPPORT VECTOR MACHINE (dengan Grid Search CV)
    svm_param_grid = {
        'C': [0.1, 1, 10, 100],
        'gamma': ['scale', 'auto', 0.1, 0.01],
        'epsilon': [0.001, 0.01, 0.1]
    }

    svm_grid = GridSearchCV(SVR(kernel='rbf'), svm_param_grid, cv=3, scoring='neg_mean_squared_error')
    svm_grid.fit(X_train, y_train)
    best_svm = svm_grid.best_estimator_

    y_pred_svm_scaled = best_svm.predict(X_test)

    # INVERSE TRANSFORM (Kembali ke Nilai Uang Asli)
    y_test_real = scaler_y.inverse_transform(y_test.reshape(-1, 1)).flatten()
    y_pred_naive = scaler_y.inverse_transform(y_pred_naive_scaled.reshape(-1, 1)).flatten()
    y_pred_kf = scaler_y.inverse_transform(y_pred_kf_scaled.reshape(-1, 1)).flatten()
    y_pred_svm = scaler_y.inverse_transform(y_pred_svm_scaled.reshape(-1, 1)).flatten()
    
    models = {
        "1. Naive Baseline": y_pred_naive,
        "2. Kalman Filter": y_pred_kf,
        "3. SVM (Optimized)": y_pred_svm
    }

    # Menyimpan rasio untuk plot nanti
    eval_results['Ratio'].append(ratio_str)

    for name, preds in models.items():
        rmse = np.sqrt(mean_squared_error(y_test_real, preds))
        mae = mean_absolute_error(y_test_real, preds)
        mape = mean_absolute_percentage_error(y_test_real, preds) * 100

        print(f"{name: <20} | RMSE: {rmse:<10.2f} | MAE: {mae:<10.2f} | MAPE: {mape:<5.2f}%")
        
        # Menyimpan nilai metrik ke dictionary
        if "Naive" in name:
            eval_results['Naive_RMSE'].append(rmse); eval_results['Naive_MAE'].append(mae); eval_results['Naive_MAPE'].append(mape)
        elif "Kalman" in name:
            eval_results['KF_RMSE'].append(rmse); eval_results['KF_MAE'].append(mae); eval_results['KF_MAPE'].append(mape)
        elif "SVM" in name:
            eval_results['SVM_RMSE'].append(rmse); eval_results['SVM_MAE'].append(mae); eval_results['SVM_MAPE'].append(mape)

    print(f"\n   [Info SVM] Parameter terbaik hasil GridSearch: C={best_svm.C}, epsilon={best_svm.epsilon}")

    # Uji Statistik Diebold-Mariano
    dm_stat, p_val = diebold_mariano_test(y_test_real, y_pred_svm, y_pred_kf)
    print(f"\n   [Uji Statistik Diebold-Mariano: SVM vs Kalman Filter]")
    print(f"   DM-Statistic = {dm_stat:.4f}, p-value = {p_val:.4f}")

    if p_val < 0.05:
        print("   -> KESIMPULAN: Performa SVM secara statistik SIGNIFIKAN lebih baik dari Kalman Filter.\n")
    else:
        print("   -> KESIMPULAN: Perbedaan performa belum cukup signifikan secara statistik (p >= 0.05).\n")

    # VISUALISASI GRAFIK PREDIKSI VS AKTUAL (Per Skenario)
    plt.figure(figsize=(10, 6))
    time_test = np.arange(1, len(y_test_real) + 1)
    
    plt.plot(time_test, y_test_real, label='Actual Profit (Test Data)', color='black', marker='o', linewidth=2)
    plt.plot(time_test, y_pred_kf, label='Kalman Filter', color='red', marker='s', linestyle='dashed', linewidth=2)
    plt.plot(time_test, y_pred_svm, label='Support Vector Machine', color='blue', marker='^', linestyle='dashed', linewidth=2)
    
    plt.title(f'Comparison of Actual vs Predicted Profit ({ratio_str})', fontsize=14, fontweight='bold')
    plt.xlabel('Testing Period (Months)', fontsize=12)
    plt.ylabel('Net Profit', fontsize=12)
    plt.legend(loc='best', fontsize=11)
    plt.grid(True, linestyle='--', alpha=0.6)
    plt.tight_layout()
    plt.show()

# ==========================================
# 6. VISUALISASI PERBANDINGAN METRIK (GRAFIK BATANG)
# ==========================================
print("\nMenampilkan Grafik Bar Perbandingan Metrik Evaluasi Keseluruhan...\n")

df_eval = pd.DataFrame(eval_results)
fig, axes = plt.subplots(1, 3, figsize=(18, 5))

ratios = df_eval['Ratio']
x = np.arange(len(ratios))  # Lokasi label di sumbu x
width = 0.25  # Lebar setiap batang

# Grafik 1: RMSE (Bar)
#axes[0].bar(x - width, df_eval['Naive_RMSE'], width, label='Naive Baseline', color='gray')
axes[0].bar(x, df_eval['KF_RMSE'], width, label='Kalman Filter', color='orange')
axes[0].bar(x + width, df_eval['SVM_RMSE'], width, label='SVM', color='blue')
axes[0].set_title('Root Mean Square Error (RMSE)')
axes[0].set_xlabel('Train-Test Split Ratio')
axes[0].set_ylabel('RMSE Value')
axes[0].set_xticks(x)
axes[0].set_xticklabels(ratios)
axes[0].legend()
axes[0].grid(True, linestyle=':', alpha=0.6, axis='y') # Grid hanya mendatar

# Grafik 2: MAE (Bar)
#axes[1].bar(x - width, df_eval['Naive_MAE'], width, label='Naive Baseline', color='gray')
axes[1].bar(x, df_eval['KF_MAE'], width, label='Kalman Filter', color='orange')
axes[1].bar(x + width, df_eval['SVM_MAE'], width, label='SVM', color='blue')
axes[1].set_title('Mean Absolute Error (MAE)')
axes[1].set_xlabel('Train-Test Split Ratio')
axes[1].set_ylabel('MAE Value')
axes[1].set_xticks(x)
axes[1].set_xticklabels(ratios)
axes[1].legend()
axes[1].grid(True, linestyle=':', alpha=0.6, axis='y')

# Grafik 3: MAPE (Bar)
#axes[2].bar(x - width, df_eval['Naive_MAPE'], width, label='Naive Baseline', color='gray')
axes[2].bar(x, df_eval['KF_MAPE'], width, label='Kalman Filter', color='orange')
axes[2].bar(x + width, df_eval['SVM_MAPE'], width, label='SVM', color='blue')
axes[2].set_title('Mean Absolute Percentage Error (MAPE)')
axes[2].set_xlabel('Train-Test Split Ratio')
axes[2].set_ylabel('MAPE (%)')
axes[2].set_xticks(x)
axes[2].set_xticklabels(ratios)
axes[2].legend()
axes[2].grid(True, linestyle=':', alpha=0.6, axis='y')

plt.tight_layout()
plt.savefig('comparison_metrics_eval_bar.png', dpi=300, bbox_inches='tight')
plt.show()
